### PR TITLE
ext_proc: Error observed when processing is STREAMED body and SEND trailer

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
Spurious unknown message logs is observed when ext_proc filter is configured with 
   request body mode: STREAMED, 
   trailer mode: SEND

The external processing stream is closed unexpectedly.

https://github.com/envoyproxy/envoy/blob/16051a60eb52ba85ae7ea6953d44a24be5a442e3/source/extensions/filters/http/ext_proc/ext_proc.cc#L679

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by yanjunxiang-google** [comment]
The root cause is that 
1) when trailers is received when the ext_proc filter is waiting for StreamedBodyCallBack state, Envoy should not send the trailer to the ext_proc server and updated the callback state. Instead it should just return and let the handleBodyResponse() call back to process the received trailer.

https://github.com/envoyproxy/envoy/blob/16051a60eb52ba85ae7ea6953d44a24be5a442e3/source/extensions/filters/http/ext_proc/ext_proc.cc#L454

Such check is already in place for other call back state. 
https://github.com/envoyproxy/envoy/blob/16051a60eb52ba85ae7ea6953d44a24be5a442e3/source/extensions/filters/http/ext_proc/ext_proc.cc#L454

Should change  it into as long as the call back state is not idle.

2) In the handleBodyResponse(), it should not send trailers if there are still pending data in the chunk_queue_ to be processed:

https://github.com/envoyproxy/envoy/blob/a420821cc2e08929fceddad30c406a4428363f6e/source/extensions/filters/http/ext_proc/processor_state.cc#L336

The trailer sending should be done when the last chunk_data response is received, at which the chunk_queue_ will be empty.

**Comment by yanjunxiang-google** [comment]
Kind Ping

**Review by yanavlasov — APPROVED** [review]

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 1]
```diff
@@ -449,8 +449,7 @@ FilterTrailersStatus Filter::onTrailers(ProcessorState& state, Http::HeaderMap&
   state.setTrailersAvailable(true);
   state.setTrailers(&trailers);
 
-  if (state.callbackState() == ProcessorState::CallbackState::HeadersCallback ||
-      state.callbackState() == ProcessorState::CallbackState::BufferedBodyCallback) {
+  if (state.callbackState() != ProcessorState::CallbackState::Idle) {
     ENVOY_LOG(trace, "Previous callback still executing -- holding header iteration");
```
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:453]: Is this log accurate any more?  we are holding more than just header iteration?

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:453]: My understanding is that trailer is also headers. I guess that's what the log means.  We can change it into "holding trailer iteration" if that is more precise.

**Pull request merged**